### ISSUE TITLE
Just implement full assignment already

### ISSUE THREAD

**Issue by masak** [masak/007#119]
This issue is the arch-enemy and complement to #118. Only one of these two can be successfully closed.

In short, this issue proposes doing basically everything #118 proposes, but without `mut`. So, `a[0][1][2] = "foo";` would just be straight-out allowed, as long as that lvalue exists. (No autovivification.)

`prefix:<++>` et al. could still be macros &mdash; and probably should be, because they're good examples of macro operators &mdash; but they can now be more straightforwardly implemented using `=`. We should probably have *some* Q abstraction corresponding to the concept of "index path". IntelliJ's PSI has some prior art there, I think.

Arrays can get `.push` and `.pop`. (And `.shift` and `.unshift` too, I guess.) Arrays and objects would both be mutable, but I don't think that matters so much. (We'll have to think about whether we want Q nodes to be mutable by default. Skirting close to #55 here.) `==` and `!=` would still compare using value semantics. You could still distinguish even equivalent arrays/objects using `.id`.

It's an open question whether `.extend` and `.update` should always produce a new clone to extend/update, or always extend/update the object itself. I think the latter makes more sense, as long as it's also easy to clone. We should have a `.clone` method.

Basically, `mut` is a cute solution in an environment where we absolutely *have* to grind the "immutable data structures" axe. But we don't in 007.

**Comment by masak** [comment]
Of course, the big hidden drawback of this proposal is that we also introduce reference sharing:

    my x = [1, 2, 3];
    my y = x;
    x[1] = "Mister Bond!";
    say(y);    # [1, "Mister Bond!", 3]

But that might be a price worth paying in order to make assignment to data structures non-excruciating. Also, we'd be in good company, among Perl, Python, Ruby, JavaScript, and most other major programming languages out there.

**Comment by masak** [comment]
A bigger sacrifice is that we let circular structures in through the door. Also not unheard of in mainstream scripting languages. But

* Stringification of arrays and objects would currently recurse infinitely on cyclical structures
* So would equality checking
* Though cyclical structures in general should be allowed, they probably shouldn't be allowed in a program Qtree. (We may not be able to prevent them in Qtrees in general, but we can make sure they never happen in the program Qtree)

**Comment by masak** [comment]
Here is a diff showing how the `hanoi.007` example would look with plain ol' assignment (and `push` and `pop`):

```diff
diff --git a/examples/hanoi.007 b/examples/hanoi.007
index 0a7aaae..9865304 100644
--- a/examples/hanoi.007
+++ b/examples/hanoi.007
@@ -13,54 +13,6 @@ my state = [
     []
 ];
 
-sub upto(n) {
-    my c = -1;
-    sub counter(irrelevant_I) {
-        c = c + 1;
-        return c;
-    }
-
-    return map(counter, [1] xx n);
-}
-
-sub pop(n) {
-    sub pop_from_nth_pile(nth) {
-        my pile = state[nth];
-        if n != nth {
-            return pile;
-        }
-
-        sub pile_item(i) { return pile[i] }
-
-        return map(pile_item, upto(elems(pile) - 1));
-    }
-
-    my L = elems(state[n]) - 1;
-    my disk = state[n][L];
-    state = map(pop_from_nth_pile, upto(3));
-    return disk;
-}
-
-sub push(disk, n) {
-    sub push_to_nth_pile(nth) {
-        my pile = state[nth];
-        if n != nth {
-            return pile;
-        }
-
-        sub pile_item(i) {
-            if i == elems(pile) {
-                return disk;
-            }
-            return pile[i];
-        }
-
-        return map(pile_item, upto(elems(pile) + 1));
-    }
-
-    state = map(push_to_nth_pile, upto(3));
-}
-
 sub tw(n, L) {
     my e = elems(state[n]);
     if L >= e {
@@ -70,7 +22,7 @@ sub tw(n, L) {
 }
 
 sub show() {
-    for reversed(upto(5)) -> L {
+    for [4, 3, 2, 1, 0] -> L {
         say(tw(0, L) ~ " " ~ tw(1, L) ~ " " ~ tw(2, L));
     }
 }
@@ -79,8 +31,8 @@ sub move(diskname, from, to) {
     say("");
     say("Moving " ~ diskname ~ " from pile " ~ str(from + 1) ~ " to pile " ~ str(to + 1) ~ "...");
     say("");
-    my disk = pop(from);
-    push(disk, to);
+    my disk = state[from].pop();
+    state[to].push(disk);
     show();
 }
 ```

**Comment by vendethiel** [comment]
Note WRT your previous comment: that'd be the same as well with the `mut` macro, right?

**Comment by masak** [comment]
> Note WRT your previous comment: that'd be the same as well with the `mut` macro, right?

I put a similar diff in [issue #118](https://github.com/masak/007/issues/118#issuecomment-170335375).

**Comment by vendethiel** [comment]
No, I meant the previous-to-last one.

**Comment by masak** [comment]
If we don't restrict the `mut` slang so as to forbid deeper things to hold shallower references, then yes, it too will permit the creation of cycles.

Restricting `mut` thus would be yet another moving part and complexity. Not restricting it would make `mut` no better than plain ol' #119 assignment. Either of those options is a strike against `mut` and a reason to prefer ordinary assignment.

**Comment by masak** [comment]
Ok, closing this one as *done*, because:

* You can now assign to array elements and object properties. (Yay!)
* The `update` and `extend` methods now modify the original object.
* Stringification and equality have been changed in such a way that they don't go into an infinite loop in the face of deeply nested references to a higher-up structure.
* The methods `.push`, `.pop`, `.shift` and `.unshift` are now implemented.

All this was a lot easier than I had feared! Wish I hadn't waited so long!
